### PR TITLE
Init: heimdall v2 basic setup

### PR DESCRIPTION
**Pull request by marcello33**
Initial setup for `heimdall v2`:z
- Deleted non-relevant files
- Removed some non-relevant packages, such as `collections`, `store`, `errors` (and many more...). Let's remember that when/if we need such things in the future, for `heimdall` purposes, we have the solution upstream.
- Implemented our `heimdall` own workflows
- Replaced `simapp` with `heimdallapp` 
- Removed `sonar`
- Setup `golangci` as linter with `heimdall` + upstream rules
- Bumped `go` version to `1.21.4`
- Solved some `govuln` issues

Note:
Do not consider the current pipeline, as it will be replaced once this PR gets merged.

### CODE REVIEW

File: docs/architecture/adr-019-protobuf-state-encoding.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ will provide two concrete implementations of the `Marshaler` interface: `AminoCo
 * `AminoCodec`: Uses Amino for both binary and JSON encoding.
 * `ProtoCodec`: Uses Protobuf for both binary and JSON encoding.
 
-Modules will use whichever codec that is instantiated in the app. By default, the Cosmos SDK's `simapp`
```
Comment (Raneet10) [inline docs/architecture/adr-019-protobuf-state-encoding.md:68]: I think we should get rid of `docs` package as a whole since the info here is catered towards Cosmos-SDK architecture rather than heimdall itself (we can add relevant stuff later ).

File: core/appmodule/README.md [hunk 2]
```diff
@@ -201,7 +201,7 @@ satisfy this dependency graph which allows staking and slashing to depend on eac
 In order to test and debug the module configuration, we need to build an app config, generally defined in a YAML file.
 This configuration should be passed first to `appconfig.LoadYAML` to get an `depinject.Config` instance.Then the
 `depinject.Config` can be passed to `depinject.Inject` and we can try to resolve dependencies in the app config.
-Alternatively, the `depinject.Config` can be created via [pure Go code](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc1/simapp/app_config.go).
```
Comment (Raneet10) [inline core/appmodule/README.md:204]: This doesn't look relevant to us.

File: docs/architecture/adr-019-protobuf-state-encoding.md (hunk 1 above)
Comment (marcello33) [reply docs/architecture/adr-019-protobuf-state-encoding.md:68]: Done https://github.com/0xPolygon/cosmos-sdk/pull/1/commits/9b7e2b5d0ede1b86b4c2aa5050a60a31af6a71a8

File: depinject/README.md [hunk 3]
```diff
@@ -75,7 +75,7 @@ Provider functions serve as the basis for the dependency tree. They are analysed
 
 `depinject` supports the use of interface types as inputs to provider functions, which helps decouple dependencies between modules. This approach is particularly useful for managing complex systems with multiple modules, such as the Cosmos SDK, where dependencies need to be flexible and maintainable.
 
-For example, `x/bank` expects an [AccountKeeper](https://pkg.go.dev/cosmossdk.io/x/bank/types#AccountKeeper) interface as [input to ProvideModule](https://github.com/cosmos/cosmos-sdk/blob/v0.47.0-rc1/x/bank/module.go#L208-L260). `SimApp` uses the implementation in `x/auth`, but the modular design allows for easy changes to the implementation if needed.
```
Comment (Raneet10) [inline depinject/README.md:78]: Same here, this isn't needed.

Comment (marcello33) [reply depinject/README.md:78]: You mean the whole READMEs or a single section within it?
Because the idea is to keep majority of the files (and most of their contents) as untouched as possible, to avoid conflict issues when performing upstream changes.
So either we delete it/them, or we keep them untouched as much as possible.
I went with the latter to keep as much info as possible for us as devs (when needed). If they don't harm the code/functionalities I'd keep them as guidelines, keeping in mind that our own stuff (as in heimdall related) will live in `heimdallapp` folder and its modules in `x/` folder. Out of that, we won't need to touch a lot, in most of the cases.
Wdyt?

File: core/appmodule/README.md (hunk 2 above)
Comment (marcello33) [reply core/appmodule/README.md:204]: Same as https://github.com/0xPolygon/cosmos-sdk/pull/1#discussion_r1407566780

File: depinject/README.md (hunk 3 above)
Comment (Raneet10) [reply depinject/README.md:78]: IMHO it'd be better to keep the whole project structure as minimal as possible. A lot of these docs/guidelines are aimed towards devs building with the SDK (as you rightly pointed), but doesn't concern Heimdall itself. To an external dev, this won't be much relevant. 

We can maybe keep note of these guidelines as they'll be available in upstream Cosmos itself for our use. What say ?

Comment (marcello33) [reply depinject/README.md:78]: So deleting all READMEs you mean right?
That would make sense instead of editing them

**Review by Raneet10 — APPROVED** [review]

File: depinject/README.md (hunk 3 above)
Comment (Raneet10) [reply depinject/README.md:78]: Yes ser!

**Review by VAIBHAVJINDAL3012 — APPROVED** [review]
